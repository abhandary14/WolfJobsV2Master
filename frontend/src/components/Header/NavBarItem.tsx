/* eslint-disable @typescript-eslint/no-explicit-any */
import { Link } from "react-router-dom";

const NavBarItem = (props: {
  link: string;
  text: string;
  setMobileMenuOpen?: any;
  mobileMenuOpen?: boolean;
}) => {
  const { link, text, setMobileMenuOpen, mobileMenuOpen } = props;
  return (
    <>
      <li onClick={() => setMobileMenuOpen(!mobileMenuOpen)}>
        <Link to={link} className="hover:text-slate-500">
          {text}
        </Link>
      </li>
    </>
  );
};

export default NavBarItem;