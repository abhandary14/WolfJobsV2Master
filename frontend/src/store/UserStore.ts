import { create } from "zustand";

type UserState = {
  name: string;
  email: string;
  password: string;
  address: string;
  role: string;
  dob: string;
  skills: string;
  phonenumber: string;
  id: string;
  availability: string;
  gender: string;
  hours: string;
  isLoggedIn: boolean;
  affiliation: string;
  resume: string;
  resumeId: string;
  unityId: string;
  studentId: string;
};

type UserAction = {
  updateName: (name: UserState["name"]) => void;
  updateEmail: (email: UserState["email"]) => void;
  updatePassword: (password: UserState["password"]) => void;
  updateAddress: (address: UserState["address"]) => void;
  updateRole: (role: UserState["role"]) => void;
  updateDob: (dob: UserState["dob"]) => void;
  updateSkills: (skills: UserState["skills"]) => void;
  updatePhonenumber: (phonenumber: UserState["phonenumber"]) => void;
  updateId: (id: UserState["id"]) => void;
  updateAvailability: (availability: UserState["availability"]) => void;
  updateGender: (gender: UserState["gender"]) => void;
  updateHours: (hours: UserState["hours"]) => void;
  updateIsLoggedIn: (isLoggedIn: UserState["isLoggedIn"]) => void;
  updateAffiliation: (affiliation: UserState["affiliation"]) => void;
  updateResume: (resume: UserState["resume"]) => void;
  updateResumeId: (resumeId: UserState["resumeId"]) => void;
  updateUnityId: (unityId: UserState["unityId"]) => void;
  updateStudentId: (studentId: UserState["studentId"]) => void;
};

export const useUserStore = create<UserState & UserAction>()((set) => ({
  name: "",
  email: "",
  password: "",
  address: "",
  role: "",
  dob: "",
  skills: "",
  phonenumber: "",
  id: "",
  availability: "",
  gender: "",
  hours: "",
  affiliation: "",
  isLoggedIn: false,
  resume: "",
  resumeId: "",
  unityId: "",
  studentId: "",

  updateName: (name: string) => {
    set(() => ({ name }));
  },
  updateEmail: (email: string) => {
    set(() => ({ email }));
  },
  updatePassword: (password: string) => {
    set(() => ({ password }));
  },
  updateAddress: (address: string) => {
    set(() => ({ address }));
  },
  updateRole: (role: string) => {
    set(() => ({ role }));
  },
  updateDob: (dob: string) => {
    set(() => ({ dob }));
  },
  updateSkills: (skills: string) => {
    set(() => ({ skills }));
  },
  updatePhonenumber: (phonenumber: string) => {
    set(() => ({ phonenumber }));
  },
  updateId: (id: string) => {
    set(() => ({ id }));
  },
  updateAvailability: (availability: string) => {
    set(() => ({ availability }));
  },
  updateGender: (gender: string) => {
    set(() => ({ gender }));
  },
  updateHours: (hours: string) => {
    set(() => ({ hours }));
  },
  updateIsLoggedIn: (isLoggedIn: boolean) => {
    set(() => ({ isLoggedIn }));
  },
  updateAffiliation: (affiliation: string) => {
    set(() => ({ affiliation }));
  },
  updateResume: (resume: string) => {
    set(() => ({ resume }));
  },
  updateResumeId: (resumeId: string) => {
    set(() => ({ resumeId }));
  },
  updateUnityId: (unityId: string) => {
    set(() => ({ unityId }));
  },
  updateStudentId: (studentId: string) => {
    set(() => ({ studentId }));
  },
}));
