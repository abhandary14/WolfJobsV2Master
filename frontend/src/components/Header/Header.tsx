import { useState } from "react";
import { useUserStore } from "../../store/UserStore";
import NavBar from "./NavBar";
import NavBarItem from "./NavBarItem";

const Header = () => {
  const isLoggedIn = useUserStore((state) => state.isLoggedIn);
  const role = useUserStore((state) => state.role);
  const [mobileMenuOpen, setMobileMenuOpen] = useState(false);

  return (
    <>
      <div className="sticky top-0 z-40 w-full bg-white backdrop-blur flex-none transition-colors duration-500 lg:z-50 lg:border-b lg:border-gray-200 supports-backdrop-blur:bg-white/95">
        <div className="max-w-8xl mx-auto">
          <div className="flex items-center justify-between py-4 border-b border-gray-200 lg:px-8 lg:border-0 mx-4 lg:mx-0">
            {/* Logo */}
            <a className="flex-none" href={isLoggedIn ? "/dashboard" : "/"}>
              <img
                alt="logo"
                src="/images/wolfjobs-logo.png"
                className="h-10"
              />
            </a>

            {/* Desktop Menu */}
            <div className="hidden lg:flex sm:items-center sm:space-x-8">
              <ul className="flex space-x-8">
                {role === "Manager" && isLoggedIn && (
                  <NavBarItem link="/dashboard" text="My Listings" />
                )}
                {role === "Applicant" && isLoggedIn && (
                  <NavBarItem link="/dashboard" text="My Applications" />
                )}
                {isLoggedIn && <NavBarItem link="/explore" text="All Jobs" />}
              </ul>
              <NavBar />
            </div>

            {/* Mobile Menu Button */}
            {isLoggedIn && (
              <div className="lg:hidden">
                <button
                  type="button"
                  className="inline-flex items-center justify-center p-2 rounded-md text-gray-500 hover:text-gray-700 hover:bg-gray-100 focus:outline-none"
                  onClick={() => setMobileMenuOpen(!mobileMenuOpen)}
                >
                  <span className="sr-only">Open main menu</span>
                  {mobileMenuOpen ? (
                    <svg
                      className="h-6 w-6"
                      xmlns="http://www.w3.org/2000/svg"
                      fill="none"
                      viewBox="0 0 24 24"
                      stroke="currentColor"
                      aria-hidden="true"
                    >
                      <path
                        strokeLinecap="round"
                        strokeLinejoin="round"
                        strokeWidth={2}
                        d="M6 18L18 6M6 6l12 12"
                      />
                    </svg>
                  ) : (
                    <svg
                      className="h-6 w-6"
                      xmlns="http://www.w3.org/2000/svg"
                      fill="none"
                      viewBox="0 0 24 24"
                      stroke="currentColor"
                      aria-hidden="true"
                    >
                      <path
                        strokeLinecap="round"
                        strokeLinejoin="round"
                        strokeWidth={2}
                        d="M4 6h16M4 12h16M4 18h16"
                      />
                    </svg>
                  )}
                </button>
              </div>
            )}
          </div>
        </div>

        {/* Mobile Menu with Animation */}
        {isLoggedIn && (
          <div
            className={`lg:hidden absolute top-[60px] w-full bg-white z-50 overflow-hidden transform transition-all duration-300 ease-in-out origin-top ${
              mobileMenuOpen ? "scale-y-100 opacity-100" : "scale-y-0 opacity-0"
            }`}
          >
            <ul className="px-2 pt-6 pb-6 space-y-3 text-center text-xl">
              {role === "Manager" && (
                <NavBarItem
                  link="/dashboard"
                  text="My Listings"
                  setMobileMenuOpen={setMobileMenuOpen}
                />
              )}
              {role === "Applicant" && (
                <NavBarItem
                  link="/dashboard"
                  text="My Applications"
                  setMobileMenuOpen={setMobileMenuOpen}
                />
              )}
              <NavBarItem
                link="/explore"
                text="All Jobs"
                setMobileMenuOpen={setMobileMenuOpen}
              />
              <NavBarItem
                link="/profile"
                text="Profile"
                setMobileMenuOpen={setMobileMenuOpen}
              />
              {role === "Applicant" && (
                <NavBarItem
                  link="/resume"
                  text="Upload Resume"
                  setMobileMenuOpen={setMobileMenuOpen}
                />
              )}
              <NavBarItem link="/logout" text="Log Out" />
            </ul>
          </div>
        )}
      </div>
    </>
  );
};

export default Header;