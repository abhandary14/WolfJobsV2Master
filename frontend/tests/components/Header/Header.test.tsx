import { render, screen } from "@testing-library/react";
import React from "react";
import Header from "../../../src/components/Header/Header";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { useUserStore } from "../../../src/store/UserStore";
import sinon from "sinon";

// Default mock user state values for missing properties
const defaultUserState = {
  name: "",
  email: "",
  password: "",
  address: "",
  dob: "",
  skills: "",
  phonenumber: "",
  id: "",
  availability: "",
  gender: "",
  hours: "",
  affiliation: "",
  resume: "",
  resumeId: "",
  unityId: "",
  studentId: "",
  isLoggedIn: false,
  role: ""
};

// Setting up mocks for user store with partial data
const setupUserStoreMock = (partialState) => {
  sinon.stub(useUserStore, "getState").returns({ ...defaultUserState, ...partialState });
};

// Helper function to render Header with routing
const renderWithRouter = (ui) => {
  return render(
    <MemoryRouter>
      <Routes>
        <Route path="*" element={ui} />
      </Routes>
    </MemoryRouter>
  );
};

describe("Header Component", () => {
  afterEach(() => {
    sinon.restore();
  });

  // Original Tests
  it("renders the logo", () => {
    renderWithRouter(<Header />);
    const logo = screen.getByAltText("logo");
    expect(logo).toBeInTheDocument(); // Use Jest's built-in matcher
    expect(logo).toHaveAttribute("src", "/images/wolfjobs-logo.png"); // Check src attribute
  });

  it("does not display 'My Listings' if user is not logged in", () => {
    setupUserStoreMock({ isLoggedIn: false });

    renderWithRouter(<Header />);

    const listingsItem = screen.queryByText("My Listings");
    expect(listingsItem).not.toBeInTheDocument(); // Use Jest's matcher for absence
  });

  // Snapshot Tests
  it("renders the Header component correctly when logged out", () => {
    setupUserStoreMock({ isLoggedIn: false });

    const { asFragment } = renderWithRouter(<Header />);
    expect(asFragment()).toMatchSnapshot(); // Snapshot test
  });

  it("renders the Header component correctly when logged in as Manager", () => {
    setupUserStoreMock({ isLoggedIn: true, role: "Manager" });

    const { asFragment } = renderWithRouter(<Header />);
    expect(asFragment()).toMatchSnapshot(); // Snapshot test
  });

  it("renders the Header component correctly when logged in as Applicant", () => {
    setupUserStoreMock({ isLoggedIn: true, role: "Applicant" });

    const { asFragment } = renderWithRouter(<Header />);
    expect(asFragment()).toMatchSnapshot(); // Snapshot test
  });

  it("renders the Header component correctly with notifications when logged in as Applicant", () => {
    setupUserStoreMock({ isLoggedIn: true, role: "Applicant" });

    const { asFragment } = renderWithRouter(<Header notificationCount={3} />);
    expect(asFragment()).toMatchSnapshot(); // Snapshot test
  });
});