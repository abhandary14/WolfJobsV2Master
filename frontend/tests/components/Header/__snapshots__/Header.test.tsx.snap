// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header Component > renders the Header component correctly when logged in as Applicant 1`] = `
<DocumentFragment>
  <div
    class="sticky top-0 z-40 w-full bg-white backdrop-blur flex-none transition-colors duration-500 lg:z-50 lg:border-b lg:border-gray-200 supports-backdrop-blur:bg-white/95"
  >
    <div
      class="max-w-8xl mx-auto"
    >
      <div
        class="flex items-center justify-between py-4 border-b border-gray-200 lg:px-8 lg:border-0 mx-4 lg:mx-0"
      >
        <a
          class="flex-none"
          href="/"
        >
          <img
            alt="logo"
            class="h-10"
            src="/images/wolfjobs-logo.png"
          />
        </a>
        <div
          class="hidden lg:flex sm:items-center sm:space-x-8"
        >
          <ul
            class="flex space-x-8"
          />
          <div
            class="relative items-center hidden ml-auto lg:flex"
          >
            <nav
              class="text-sm font-semibold leading-6 text-slate-700 "
            >
              <ul
                class="flex space-x-8"
              />
            </nav>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Header Component > renders the Header component correctly when logged in as Manager 1`] = `
<DocumentFragment>
  <div
    class="sticky top-0 z-40 w-full bg-white backdrop-blur flex-none transition-colors duration-500 lg:z-50 lg:border-b lg:border-gray-200 supports-backdrop-blur:bg-white/95"
  >
    <div
      class="max-w-8xl mx-auto"
    >
      <div
        class="flex items-center justify-between py-4 border-b border-gray-200 lg:px-8 lg:border-0 mx-4 lg:mx-0"
      >
        <a
          class="flex-none"
          href="/"
        >
          <img
            alt="logo"
            class="h-10"
            src="/images/wolfjobs-logo.png"
          />
        </a>
        <div
          class="hidden lg:flex sm:items-center sm:space-x-8"
        >
          <ul
            class="flex space-x-8"
          />
          <div
            class="relative items-center hidden ml-auto lg:flex"
          >
            <nav
              class="text-sm font-semibold leading-6 text-slate-700 "
            >
              <ul
                class="flex space-x-8"
              />
            </nav>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Header Component > renders the Header component correctly when logged out 1`] = `
<DocumentFragment>
  <div
    class="sticky top-0 z-40 w-full bg-white backdrop-blur flex-none transition-colors duration-500 lg:z-50 lg:border-b lg:border-gray-200 supports-backdrop-blur:bg-white/95"
  >
    <div
      class="max-w-8xl mx-auto"
    >
      <div
        class="flex items-center justify-between py-4 border-b border-gray-200 lg:px-8 lg:border-0 mx-4 lg:mx-0"
      >
        <a
          class="flex-none"
          href="/"
        >
          <img
            alt="logo"
            class="h-10"
            src="/images/wolfjobs-logo.png"
          />
        </a>
        <div
          class="hidden lg:flex sm:items-center sm:space-x-8"
        >
          <ul
            class="flex space-x-8"
          />
          <div
            class="relative items-center hidden ml-auto lg:flex"
          >
            <nav
              class="text-sm font-semibold leading-6 text-slate-700 "
            >
              <ul
                class="flex space-x-8"
              />
            </nav>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Header Component > renders the Header component correctly with notifications when logged in as Applicant 1`] = `
<DocumentFragment>
  <div
    class="sticky top-0 z-40 w-full bg-white backdrop-blur flex-none transition-colors duration-500 lg:z-50 lg:border-b lg:border-gray-200 supports-backdrop-blur:bg-white/95"
  >
    <div
      class="max-w-8xl mx-auto"
    >
      <div
        class="flex items-center justify-between py-4 border-b border-gray-200 lg:px-8 lg:border-0 mx-4 lg:mx-0"
      >
        <a
          class="flex-none"
          href="/"
        >
          <img
            alt="logo"
            class="h-10"
            src="/images/wolfjobs-logo.png"
          />
        </a>
        <div
          class="hidden lg:flex sm:items-center sm:space-x-8"
        >
          <ul
            class="flex space-x-8"
          />
          <div
            class="relative items-center hidden ml-auto lg:flex"
          >
            <nav
              class="text-sm font-semibold leading-6 text-slate-700 "
            >
              <ul
                class="flex space-x-8"
              />
            </nav>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
